man/man1/restart2yaml.1
.TH restart2yaml 1  "USER COMMANDS"
.SH NAME
restart2yaml \- Convert LPJmL restart file into YAML/JSON.
.SH SYNOPSIS
.B restart2yaml
[\-h] [\-name key] [\-first] [\-json] [\-noindent] [\-invalid2null] [\-d n] [\-t] \fIrestartfile\fP [start [end]]

.B restart2json
[\-name key] [\-first] [\-noindent] [\-invalid2null] [\-t] \fIrestartfile\fP [start [end]]
.SH DESCRIPTION
Program prints contents of restart file of a LPJmL model run in YAML/JSON format. \fBrestart2json\fP is an alias for \fBrestart2yaml\fP -json.
.SH OPTIONS
.TP
\-h,\--help
Print a short help text.
.TP
\-name key
Print only variables/array/structures with name key.
.TP
\-first
Stop at first occurrence of key specified by the \-name option.
.TP
\-json
Output format is set to JSON, default is YAML.
.TP
\-noindent
Disable indentations in JSON files to save storage space, default is 2 spaces per level.
.TP
\-invalid2null
Set all not a number (NaN) and infinite values to null. This is in particular useful for JSON output format, because these values are not defined in the JSON standard.
.TP
\-d n
Set the number of decimal places for float/double values, default is 16.
.TP
\-t
Print only table of defined object names.
.TP
.I restartfile
Name of restart file.
.TP
.I start
Index of first grid cell to be printed.
.TP
.I end
Index of last grid cell to be printed.

.SH EXAMPLE
.TP
Print restart file:
.B restart2yaml
restart.lpj >restart.yaml
.PP
.TP
Print restart file as JSON file without indentation:
.B restart2yaml
-json -noindent restart.lpj >restart.json
.PP
.SH EXIT STATUS
.B
restart2yaml
returns a zero exit status if the restart file has been printed.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
json2restart(1), cmprestart(1), lpjprint(1), lpjml(1), lpjcheck(1), lpjfiles(1)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security


## [Unreleased]


## [6.1.4] - 2026-07-03

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Fabian Stenzel (stenzel@pik-potsdam.de), David Hötten (davidho@pik-potsdam.de)

### Added

- Utility `cmprestart` added to compare two restart files and check for identity.
- Function `bstruct_skiparray()` and corresponding test `test_skiparray.c` added in `bstruct` library to skip arrary elements.
- Function `bstruct_gettype()` added to `bstruct` library to get datatype of object in restart file.
- Function `bstrucht_printnamestack()` added for more comprehensive error messages.
- Comparison of restart files with `cmprestart` added to merge request template.

### Changed

- Index array object is always skipped in `bstruct_findobject.c`, test `test_indexarray.c` updated to test for reading array without seeking to an index array.
- Check for seeking to position within file added to `bstruct_findobject.c`, `bstruct_open.c`, `bstruct_wopen.c` and `bstruct_seekindexarray.c`.
- Constant `MAXLEVEL` renamed to `BSTRUCT_MAXLEVEL`.
- Macros `getname()` and `isinvalidtoken()` renamed to `bstruct_getname()` and `bstruct_isinvalidtoken()`.
- String comparison in `bstruct_fprintdata` replaced by call to `isnan()` and `isinf()`.
- New token `bstruct_dzero` defined for restart files to distinguish between float and double zeros, version of bstruct files changed to 2.


## [6.1.3] - 2026-07-03

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth(wirth@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Jannes Breier (breier@pik-potsdam.de)

### Added

- Option `-metafile` and `-json` added to `cutclm` and `catclm`.
- Option `-double` added to `cdf2soil.c` to write grid file with double datatype.

### Changed

- New datatype `Metadata` defined in `types.h` combining maps, attributes, basetemp and hlimit arrays used by `opendata()`.
- basetemp, countrymap and hlimit arrays are copied from JSON metafiles to newly created metafiles in `regridclm`, `cutlcm`, `catclm`, and `mathclm`.
- Scaling to sum of stand fractions removed for output `"vegc_avg"`.
- Utility `getcountry` reads now JSON metafile for country codes including the alpha-3 country codes and global grid instead of CLM file.
- If `-pedantic` option is set, non-matching titles in input files stops `lpjml` with an error.
- Datatype of `lon`, `lat`, and `area` elements of `Coord` changed from `Real` to `double`.
- Missing path to grid file added in `openmetafile.c`.

### Fixed

- Warning for float coordinates corrected in `cdf2coord.c`, `cdf2grid.c` and `cdf2soil.c`.
- Variables initialized to zero in `reservoir2cdf.c` and `restart2yaml.c` to avoid warnings by `gcc`.
- Carbon balance fixed in `reclaim_land.c`.
- Check for NULL pointer added in `getuser.c` to avoid SEGV if user name cannot be resolved.
- Verbose output corrected for input without time axis in `cdf2clm.c`.
- Output is rescaled by the actual sum of stand fractions in `update_daily_cell.c` and `fwriteoutput.c` to avoid division by zero (issue #411).
- Default timestep set to 1 in `regridclm.c`.
- Variable `depth` initialized to zero before layer loop in `fwriteoutput.c` to correct `soilc_1m` output.
- Length of buffer for hostname corrected in `gethost.c`.
- Output of unprintable characters corrected in `fputprintable.c`.
- Memory leak fixed in `openmetafile.c` in case of error.


## [6.1.2] - 2026-06-16

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Fixed

- Check for last year for lightning and human ignition disabled in `filesexist.c`, `lpjcheck` does not return an error for standard input.


## [6.1.1] - 2026-06-12

### Contributors

author: Jannes Breier (breier@pik-potsdam.de),  Werner von Bloh (bloh@pik-potsdam.de)
code review: Christoph Müller (cmueller@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)

### Added

- `logo.png` added to `man` directory
- `LICENSES` folder
- `LICENSES/CC0-1.0.md` added for artwork license

### Changed

- `AUTHORS` renamed to `AUTHORS.md` and restructured for active and previous authors
- `INSTALL` renamed to `INSTALL.md` and adjusted
- `LICENSE` renamed to `LICENSES/AGPL-3.0.md` and moved to `LICENSES` folder
- `README` renamed to `README.md` and expanded
- `configure.sh`, `src/lpj/copyright.c` and `src/lpj/printlicense.c` adjusted to reflect renaming of files
- Updated module version for `oneAPI` in `run_simulate_default.sh`


## [6.1.0] - 2026-06-12

### Contributors

- author: Luke Oberhagemann (lukeober@pik-potsdam.de), Markus Drüke (Markus.Drueke@dwd.de), Maik Billing (billing@pik-potsdam.de), Werner von Bloh (bloh@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Added

- Option `-inpath` added to `configure.sh` to set path for input files.
- Function `fprintclimbuf()` added in `climbuf.c` to print variables and is called in `fprintcell.c`.
- Function `fprintbuffer()` added in `buffer.c` to print variables and is called in `fprintclimbuf.c`.
- Lakes setting stored in restart file and setting compared to actual setting in `openrestart.c`.

src/utils/restart2yaml.c
/**************************************************************************************/
/**                                                                                \n**/
/**                   r  e  s  t  a  r  t  2  y  a  m  l  .  c                     \n**/
/**                                                                                \n**/
/**     C implementation of LPJmL                                                  \n**/
/**                                                                                \n**/
/**     Utility converts restart file into yaml or JSON file                       \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <math.h>
#include "types.h"
#include "errmsg.h"
#include "swap.h"
#include "hash.h"
#include "bstruct.h"

#define USAGE "Usage: %s [-h] [-name key] [-first] [-json] [-noindent] [-invalid2null] [-d n]\n"\
              "       [-t] restartfile [first [last]]\n"
#define ERR_USAGE USAGE "\nTry \"%s --help\" for more information.\n"

static void printname(const char *name)
{
  /* to check for string N is necessary because N will be interpreted as FALSE in YAML 1.1 */
  fputprintable(stdout,name);
  fputs(":",stdout);
} /* of 'printname' */

int main(int argc,char **argv)
{
  Bstruct file;
  Bstruct_data data;
  const char *progname;
  char *endptr;
  char *key=NULL;
  int level,iarg,keylevel;
  Bool notend,isarray=FALSE,iskey=TRUE,stop=FALSE,setnull=FALSE;
  int firstcell,lastcell=0,last=0,cell,decimals=16,rc;
  Bool first=FALSE,islastcell=FALSE,isjson=FALSE,indent=TRUE,istable=FALSE;
  progname=strippath(argv[0]);
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-h") || !strcmp(argv[iarg],"--help"))
      {   
        fputs("     ",stdout);
        rc=printf("%s (" __DATE__ ") Help",
                  progname);
        fputs("\n     ",stdout);
        repeatch('=',rc);
        printf("\n\nConvert restart file into YAML/JSON file for LPJmL %s\n\n",getversion());
        printf(USAGE,progname);
        printf("\nArguments:\n"
               "-h,--help     print this help text\n"
               "-name key     print only variables/array/structures with name key\n"
               "-first        stop at first occurrence of key specified by the -name option\n"
               "-json         output format is set to JSON, default is YAML\n"
               "-noindent     disable indentations in JSON files, default is 2 spaces per level\n"
               "-invalid2null set all NaN and infinite values to null\n"
               "-d n          set the number of decimal places for float/double values, default is 16\n"
               "-t            print only table of defined object names\n"
               "restartfile   name of restart file\n"
               "start         index of first grid cell to be printed\n"
               "end           index of last grid cell to be printed\n\n"
               "(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file\n");
        return EXIT_SUCCESS;
      }   
      else if(!strcmp(argv[iarg],"-name"))
      {
        if(argc==iarg+1)
        {
          fprintf(stderr,"Missing argument after option '-name'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        key=argv[++iarg];
        iskey=FALSE;
      }
      else if(!strcmp(argv[iarg],"-d"))
      {
        if(iarg==argc-1)
        {
          fprintf(stderr,"Missing argument after option '-d'.\n"
                  ERR_USAGE,progname,progname);
          return EXIT_FAILURE;
        }
        decimals=strtol(argv[++iarg],&endptr,10);
        if(*endptr!='\0')
        {
          fprintf(stderr,"Invalid value '%s' for option '-d'.\n",
                  argv[iarg]);
          return EXIT_FAILURE;
        }
        if(decimals<1)
        {
          fprintf(stderr,"Number of decimals=%d must be greater than zero.\n",decimals);
          return EXIT_FAILURE;
        }
      }
      else if(!strcmp(argv[iarg],"-json"))
        isjson=TRUE;
      else if(!strcmp(argv[iarg],"-invalid2null"))
        setnull=TRUE;
      else if(!strcmp(argv[iarg],"-noindent"))
        indent=FALSE;
      else if(!strcmp(argv[iarg],"-t"))
        istable=TRUE;
      else if(!strcmp(argv[iarg],"-first"))
        stop=TRUE;
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                ERR_USAGE,argv[iarg],progname,progname);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<iarg+1)
  {
    fprintf(stderr,"Missing argument(s).\n"
            ERR_USAGE,progname,progname);
    return EXIT_FAILURE;
  }
  file=bstruct_open(argv[iarg],TRUE);
  if(file==NULL)
    return EXIT_FAILURE;
  if(istable)
  {
    if(isjson)
    {
      printf("{\n\"filename\" : \"%s\",\n",argv[iarg]);
      bstruct_printnametable("table",file,TRUE);
      puts("\n}");
    }
    else
    {
      printf("%%YAML 1.2\n"
             "---\n"
             "filename: \"%s\"\n",
             argv[iarg]);
      bstruct_printnametable("table",file,FALSE);
      puts("...");
    }
    bstruct_finish(file);
    return EXIT_SUCCESS;
  }
  firstcell=0;
  if(argc>iarg+1)
  {
    firstcell=strtol(argv[iarg+1],&endptr,10);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for first grid.\n",argv[iarg+1]);
      return EXIT_FAILURE;
    }
    if(firstcell<0)
    {
      fprintf(stderr,"First cell must be >=0.\n");
      return EXIT_FAILURE;
    }
  }
  if(argc>iarg+2)
  {
    lastcell=strtol(argv[iarg+2],&endptr,10);
    if(*endptr!='\0')
    {
      fprintf(stderr,"Invalid number '%s' for last grid.\n",argv[iarg+2]);
      return EXIT_FAILURE;
    }
    islastcell=TRUE;
  }
  if(isjson)
    printf("{\n\"filename\" : \"%s\"",argv[iarg]);
  else
    printf("%%YAML 1.2\n"
           "---\n"
           "filename: \"%s\"\n",
           argv[iarg]);
  level=0;
  cell=firstcell;
  notend=TRUE;
  keylevel=0;
  while(notend)
  {
    if(bstruct_readdata(file,&data))
    {
      fprintf(stderr,"Error reading cell %d in '%s'.\n",cell,argv[iarg]);
      break;
    }
#ifdef DEBUG
    printf("%d,%d,%s:",cell,level,bstruct_typenames[data.token]);
#endif
    switch(data.token)
    {
      case BSTRUCT_END:
        notend=FALSE;
        break;
      case BSTRUCT_ENDSTRUCT: case BSTRUCT_ENDARRAY:
        level--;
        if(isjson && iskey)
        {
          putchar('\n');
          if(indent)
            repeatch(' ',2*(level-keylevel));
          putchar(data.token==BSTRUCT_ENDSTRUCT ? '}' : ']');
           first=FALSE;
        }
        if(key!=NULL && level==keylevel)
        {
          if(iskey && stop)
            notend=FALSE;
          iskey=FALSE;
        }
        /* if level reached 1 data for one cell has been read completely, increase cell counter */
        if(isarray && level==1)
        {
          cell++;
          if(cell>last)
            notend=FALSE;
        }
        break;
      case BSTRUCT_BEGINSTRUCT: case BSTRUCT_BEGINARRAY: case BSTRUCT_BEGINARRAY1:
        if(key!=NULL && data.name!=NULL && !strcmp(data.name,key))
        {
          iskey=TRUE;
          keylevel=level;
        }
        if(iskey)
        {
          if(isjson)
          {
            if(!first)
              puts(",");
            first=TRUE;
            if(indent)
              repeatch(' ',2*(level-keylevel));
            if(bstruct_isunnamed(&data))
              puts(data.token==BSTRUCT_BEGINSTRUCT ? "{" : "[");
            else
              printf("\"%s\" : %c\n",data.name,data.token==BSTRUCT_BEGINSTRUCT ? '{' : '[');
          }
          else
          {
            if(first)
              first=FALSE;
            else
              repeatch(' ',2*(level-keylevel));
            if(bstruct_isunnamed(&data))
            {
              fputs("- ",stdout);
              first=TRUE;
            }
            else
            {
              printname(data.name);
              if(data.token!=BSTRUCT_BEGINSTRUCT && data.size==0)
                puts(" []");
              else
                fputc('\n',stdout);
            }
          }
        }
        level++;
        break;
      case BSTRUCT_INDEXARRAY:
        isarray=TRUE;
        if(islastcell)
          last=lastcell;
        else
          last=data.size-1;
        if(last>=data.size)
        {
          fprintf(stderr,"Last cell %d is greater than upper number of cells %d.\n",last,data.size-1);
          bstruct_finish(file);
          return EXIT_FAILURE;
        }
        if(firstcell>=data.size)
        {
          fprintf(stderr,"First cell %d is greater than upper number of cells %d.\n",firstcell,data.size-1);
          bstruct_finish(file);
          return EXIT_FAILURE;
        }
        fseek(bstruct_getfile(file),data.data.index[firstcell],SEEK_SET);
        break;
      default:
        if(key!=NULL && data.name!=NULL && !strcmp(data.name,key))
        {
          if(stop)
            notend=FALSE;
          iskey=TRUE;
          keylevel=level;
        }
        if(!iskey)
          break;
        if(isjson)
        {
          if(first)
            first=FALSE;
          else
            puts(",");
          if(indent)
            repeatch(' ',2*(level-keylevel));
          if(data.name!=NULL)
            printf("\"%s\" : ",data.name);
        }
        else
        {
          if(first)
            first=FALSE;
          else
            repeatch(' ',2*(level-keylevel));
          if(bstruct_isunnamed(&data))
            fputs("- ",stdout);
          else
          {
            printname(data.name);
            putchar(' ');
          }
        }
        if(data.token==BSTRUCT_FLOAT)
        {
          if(isnan(data.data.f))
          {
            fprintf(stderr,"Error: Value of '%s' is not a number (Nan).\n",bstruct_getname(data.name));
            if(setnull)
              data.token=BSTRUCT_NULL;
          }
          else if(isinf(data.data.f))
          {
            fprintf(stderr,"Error: Value of '%s' is infinite.\n",bstruct_getname(data.name));
            if(setnull)
              data.token=BSTRUCT_NULL;
          }
        }
        else if(data.token==BSTRUCT_DOUBLE)
        {
          if(isnan(data.data.d))
          {
            fprintf(stderr,"Error: Value of '%s' is not a number (Nan).\n",bstruct_getname(data.name));
            if(setnull)
              data.token=BSTRUCT_NULL;
          }
          else if(isinf(data.data.d))
          {
            fprintf(stderr,"Error: Value of '%s' is infinite.\n",bstruct_getname(data.name));
            if(setnull)
              data.token=BSTRUCT_NULL;
          }
        }
        bstruct_printdata(&data,decimals);
        if(!isjson)
          fputc('\n',stdout);
        if(key!=NULL && data.name!=NULL && !strcmp(data.name,key))
          iskey=FALSE;
        break;
    } /* of switch(token) */
    bstruct_freedata(&data);
  } /* of while */
  if(isjson)
  {
    if(data.token==BSTRUCT_END)
      puts("\n}");
    else
    {
      if(key==NULL)
        puts("\n]");
      puts("}");
    }
  }
  else
    puts("...");
  bstruct_finish(file);
  return EXIT_SUCCESS;
} /* of 'main' */
